core/http/src/header/accept.rs

```diff
@@ -147,24 +147,8 @@ impl Accept {
         let mut all = self.iter();
         let mut preferred = all.next().unwrap_or(&ANY);
         for media_type in all {
-            if media_type.weight().is_none() && preferred.weight().is_some() {
-                // Media types without a `q` parameter are preferred.
+            if media_type.better_than(preferred) {
                 preferred = media_type;
-            } else if media_type.weight_or(0.0) > preferred.weight_or(1.0) {
-                // Prefer media types with a greater weight, but if one doesn't
-                // have a weight, prefer the one we already have.
-                preferred = media_type;
-            } else if media_type.specificity() > preferred.specificity() {
-                // Prefer more specific media types over less specific ones. IE:
-                // text/html over application/*.
-                preferred = media_type;
-            } else if media_type == preferred {
-                // Finally, all other things being equal, prefer a media type
-                // with more parameters over one with fewer. IE: text/html; a=b
-                // over text/html.
-                if media_type.params().count() > preferred.params().count() {
```

Comment:
This `if` can not be entered, because at this point `media_type == preferred`.

I assumed that this is intended to be a kind of lexicography comparator (check several conditions, and each one is strictly more important than the next) when I rewrote it. Please correct me if I misunderstood the intended algorithm.

Reply:
The first equality check does not consider parameters, as documented. Please don't change this method at all.

Reply:
I think I'll roll back this change for now and discuss it in other PR, to make review easier.